[code] : Re-adding parent relationship after removal throws an error

Reported via Support: [103454](https://dotcms.zendesk.com/agent/tickets/103454)
Reproducible as of dotCMS 5.2.8.3, and latest master.

The parent side of a Relationship between two Content Types can be deleted for any reason. Now, if the user decides to bring it back, an error is being displayed on both the UI and the log.

**To Reproduce**
Steps to reproduce the behavior:
1. Create two test Content Types named `Parent CT`, and `Child CT`
2. Go to the `Parent CT`, and add a **Relationships Field** with the following values:
2.1. Name = `Child Rel`
2.2. `New`
2.3. Content Type = `Child CT`
2.4. Relationship = `Many to many`
2.5. `Save`
3. Go to the `Child CT` this time, and add a **Relationships Field** with the following values:
3.1. Name = `Parent Rel`
3.2. `Existing`
3.3. Relationship = `ParentCt.childRel.Mano to many`
3.4. `Save`
4. Go back to the `Parent CT`, and DELETE the `Child Rel` field.
5. Drag-and-drop a new Relationships field again to bring the previous relationship back, using the following values:
5.1. Name = `Child Rel`
5.2. `Existing`
5.3. Relationship = `ParentCt.childRel.Mano to many`
5.4. `Save`
6. The following error shows up in the UI:
```
Bad request error

Unfortunately there was an error. Please try again, and if the error persists, please contact support.
```
And this error is logged to the `dotcms.log` file:
```
Caused by: com.dotcms.contenttype.exception.NotFoundInDbException: Field variable with var:parentRel not found
at com.dotcms.contenttype.business.FieldFactoryImpl.byContentTypeFieldVar(FieldFactoryImpl.java:64) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.contenttype.business.FieldAPIImpl.byContentTypeAndVar_aroundBody14(FieldAPIImpl.java:733) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.contenttype.business.FieldAPIImpl$AjcClosure15.run(FieldAPIImpl.java:1) ~[dotcms_21.01_999999.jar:?]
at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.8.10.jar:?]
at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.contenttype.business.FieldAPIImpl.byContentTypeAndVar(FieldAPIImpl.java:732) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.contenttype.business.FieldAPIImpl.updateRelationshipObject(FieldAPIImpl.java:481) ~[dotcms_21.01_999999.jar:?]
at com.dotcms.contenttype.business.FieldAPIImpl.getRelationshipForField(FieldAPIImpl.java:385) ~[dotcms_21.01_999999.jar:?]
... 117 more
```

**Expected behavior**
The application needs to allow users to re-add the parent side of a Relationship after its deletion.

--- a/dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java
+++ b/dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java
@@ -1528,6 +1528,69 @@ public class FieldAPITest extends IntegrationTestBase {
         }
     }
 
+    /**
+     * Method to test: com.dotcms.contenttype.business.FieldAPI#save(com.dotcms.contenttype.model.field.Field, com.liferay.portal.model.User, boolean)
+     * Given scenario: Recreate a deleted relationship field but choosing exiting relationship, from parent to child
+     * Expected result: Should succeed
+     */
+
+    @Test
+    public void testSaveRelationshipField_GivenRecreatedDeletedRelationshipFieldChoosingExistingRel_SaveShouldSucceed()
+            throws DotSecurityException, DotDataException {
+        final long time = System.currentTimeMillis();
+
+        ContentType parentContentType = null;
+        ContentType childContentType  = null;
+
+        try {
+            parentContentType = createAndSaveSimpleContentType("parentContentType" + time);
+            childContentType = createAndSaveSimpleContentType("childContentType" + time);
+
+            //Adding a RelationshipField to the parent
+            final Field parentTypeRelationshipField = createAndSaveRelationshipField("newRel",
+                    parentContentType.id(), childContentType.variable(), CARDINALITY);
+
+            final String fullFieldVar =
+                    parentContentType.variable() + StringPool.PERIOD + parentTypeRelationshipField.variable();
+
+            //Adding a RelationshipField to the child
+            final Field childTypeRelationshipField = createAndSaveRelationshipField("otherSideRel",
+                    childContentType.id(), fullFieldVar, CARDINALITY);
+
+            final Relationship relationship = relationshipAPI.byTypeValue(fullFieldVar);
+
+            assertNotNull(relationship);
+
+            assertEquals(childContentType.id(), relationship.getChildStructureInode());
+            assertEquals(parentTypeRelationshipField.variable(), relationship.getChildRelationName());
+            assertEquals(parentContentType.id(), relationship.getParentStructureInode());
+            assertEquals(childTypeRelationshipField.variable(), relationship.getParentRelationName());
+            assertEquals(CARDINALITY, Integer.toString(relationship.getCardinality()));
+            assertEquals(fullFieldVar, relationship.getRelationTypeValue());
+
+            final String parentTypeRelationshipFieldVariable = parentTypeRelationshipField.variable();
+            // let's delete the relationship field from the parent content type
+            APILocator.getContentTypeFieldAPI().delete(parentTypeRelationshipField);
+
+            // let's attempt to re-add the field but choosing the existing relationship
+            final Field recreatedParentTypeRelationshipField = createAndSaveRelationshipField("newRel",
+                    parentContentType.id(), parentContentType.variable() + "." + parentTypeRelationshipFieldVariable
+                    , CARDINALITY);
+
+            assertEquals(recreatedParentTypeRelationshipField.variable(),
+                    relationship.getChildRelationName());
+
+        } finally {
+            if (UtilMethods.isSet(parentContentType) && UtilMethods.isSet(parentContentType.id())) {
+                contentTypeAPI.delete(parentContentType);
+            }
+
+            if (UtilMethods.isSet(childContentType) && UtilMethods.isSet(childContentType.id())) {
+                contentTypeAPI.delete(childContentType);
+            }
+        }
+    }
+
     static class GraphQLFieldNameCompatibilityTestCase {
         String fieldName;
         private Class<? extends Field> fieldType;

--- a/dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java
+++ b/dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java
@@ -1530,7 +1530,7 @@ public class FieldAPITest extends IntegrationTestBase {
 
     /**
      * Method to test: com.dotcms.contenttype.business.FieldAPI#save(com.dotcms.contenttype.model.field.Field, com.liferay.portal.model.User, boolean)
-     * Given scenario: Recreate a deleted relationship field but choosing exiting relationship, from parent to child
+     * Given scenario: Recreate a deleted relationship field but choosing existing relationship, from parent to child
      * Expected result: Should succeed
      */
 
@@ -1571,6 +1571,9 @@ public class FieldAPITest extends IntegrationTestBase {
             final String parentTypeRelationshipFieldVariable = parentTypeRelationshipField.variable();
             // let's delete the relationship field from the parent content type
             APILocator.getContentTypeFieldAPI().delete(parentTypeRelationshipField);
+
+            final Relationship reFetchedRelationship = relationshipAPI.byTypeValue(fullFieldVar);
+            assertNull(reFetchedRelationship.getChildRelationName());
 
             // let's attempt to re-add the field but choosing the existing relationship
             final Field recreatedParentTypeRelationshipField = createAndSaveRelationshipField("newRel",

--- a/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
@@ -80,6 +80,7 @@ import com.liferay.portal.language.LanguageException;
 import com.liferay.portal.language.LanguageUtil;
 import com.liferay.portal.model.User;
 
+import io.vavr.control.Try;
 import java.net.ConnectException;
 import java.util.*;
 
@@ -480,10 +481,10 @@ public class FieldAPIImpl implements FieldAPI {
 
             //verify if the cardinality was changed to update it on the other side of the relationship
             if (relationship.getParentRelationName() != null) {
-                final Field otherSideField = byContentTypeAndVar(relatedContentType,
-                        relationship.getParentRelationName());
+                final Field otherSideField = Try.of(()->byContentTypeAndVar(relatedContentType,
+                        relationship.getParentRelationName())).getOrNull();
 
-                if (!otherSideField.values().equals(field.values())) {
+                if (otherSideField!=null && !otherSideField.values().equals(field.values())) {
                     //if cardinality changes, the other side field will be updated with the new cardinality
                     builder = FieldBuilder.builder(otherSideField);
                     fieldFactory.save(builder.values(field.values()).build());

--- a/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
@@ -478,19 +478,7 @@ public class FieldAPIImpl implements FieldAPI {
                 builder.required(false);
                 fieldFactory.save(builder.values(field.values()).build());
             }
-
-            //verify if the cardinality was changed to update it on the other side of the relationship
-            //todo remove IF
-//            if (relationship.getParentRelationName() != null) {
-                final Field otherSideField = Try.of(()->byContentTypeAndVar(relatedContentType,
-                        relationship.getParentRelationName())).getOrNull();
-
-                if (otherSideField!=null && !otherSideField.values().equals(field.values())) {
-                    //if cardinality changes, the other side field will be updated with the new cardinality
-                    builder = FieldBuilder.builder(otherSideField);
-                    fieldFactory.save(builder.values(field.values()).build());
-                }
-//            }
+            
         }
         relationship.setCardinality(cardinality);
     }

--- a/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
@@ -490,6 +490,7 @@ public class FieldAPIImpl implements FieldAPI {
                     fieldFactory.save(builder.values(field.values()).build());
                 }
             }
+
         }
         relationship.setCardinality(cardinality);
     }

--- a/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
@@ -451,18 +451,6 @@ public class FieldAPIImpl implements FieldAPI {
                 builder.required(false);
                 fieldFactory.save(builder.build());
             }
-
-            if (relationship.getChildRelationName() != null) {
-                //verify if the cardinality was changed to update it on the other side of the relationship
-                final Field otherSideField = byContentTypeAndVar(relatedContentType,
-                        relationship.getChildRelationName());
-
-                if (!otherSideField.values().equals(field.values())) {
-                    //if cardinality changes, the other side field will be updated with the new cardinality
-                    builder = FieldBuilder.builder(otherSideField);
-                    fieldFactory.save(builder.values(field.values()).build());
-                }
-            }
         } else {
             //child is updated
             relationship.setChildRelationName(relationName);
@@ -478,7 +466,7 @@ public class FieldAPIImpl implements FieldAPI {
                 builder.required(false);
                 fieldFactory.save(builder.values(field.values()).build());
             }
-            
+
         }
         relationship.setCardinality(cardinality);
     }

--- a/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java
@@ -480,7 +480,8 @@ public class FieldAPIImpl implements FieldAPI {
             }
 
             //verify if the cardinality was changed to update it on the other side of the relationship
-            if (relationship.getParentRelationName() != null) {
+            //todo remove IF
+//            if (relationship.getParentRelationName() != null) {
                 final Field otherSideField = Try.of(()->byContentTypeAndVar(relatedContentType,
                         relationship.getParentRelationName())).getOrNull();
 
@@ -489,7 +490,7 @@ public class FieldAPIImpl implements FieldAPI {
                     builder = FieldBuilder.builder(otherSideField);
                     fieldFactory.save(builder.values(field.values()).build());
                 }
-            }
+//            }
         }
         relationship.setCardinality(cardinality);
     }

